README.md

```diff
@@ -1,6 +1,80 @@
+
+Coroutines should be the preferred way of writing asynchronous code, as it allows the use of `try` / `catch` and doesn't result in a so-called callback hell. Coroutines do not use the returned `Promise` of a `then` callback, but returning a `Promise` is required by Promises/A+. Thus there's a lot of useless object creations, which aren't cheap and add up.
+
+This specification does not deal with how to create, succeed or fail `Awaitable`s, as only the consumption of `Awaitable`s is required to be interoperable.
+
+## Terminology
+
+1. _Awaitable_ is an object implementing `Interop\Async\Awaitable` and conforming to this specification.
+1. _Value_ is any legal PHP value (including `null`), except an instance of `Interop\Async\Awaitable`.
+1. _Error_ is any value that can be thrown using the `throw` statement.
+1. _Reason_ is an error indicating why an `Awaitable` has failed.
+
+## States
+
+An `Awaitable` MUST be in one of three states: `pending`, `succeeded`, `failed`.
+
+| A promise in … state | &nbsp; |
+|----------------------|--------|
+|`pending`  | <ul><li>MAY transition to either the `succeeded` or `failed` state.</li></ul>                                |
+|`succeeded`| <ul><li>MUST NOT transition to any other state.</li><li>MUST have a value which MUST NOT change.*</li></ul>  |
+|`failed`   | <ul><li>MUST NOT transition to any other state.</li><li>MUST have a reason which MUST NOT change.*</li></ul> |
+
+* _Must not change_ refers to the _reference_ being immutable in case of an object, _not the object itself_ being immutable.
+
+## Consumption
+
+An `Awaitable` MUST implement `Interop\Async\Awaitable` and thus provide a `when` method to access its current or eventual value or reason.
+
+```php
+<?php
+
+namespace Interop\Async;
+
+/**
+ * Representation of a the future value of an asynchronous operation.
+ */
+interface Awaitable
+{
+    /**
+     * Registers a callback to be invoked when the awaitable is resolved.
+     *
+     * @param callable(\Throwable|\Exception|null $exception, mixed $result) $onResolved
+     *
+     * @return void
+     */
+    public function when(callable $onResolved);
+}
+```
+
+The `when` method MUST accept at least one argument:
+
+`$callback` – A callable conforming to the following signature:
+
+```php
+function($error, $value) { /* ... */ }
+```
+
+Any implementation MUST at least provide these two parameters. The implementation MAY extend the `Awaitable` interface with additional parameters passed to the callback. Further arguments to `when` MUST have default values, so `when` can always be called with only one argument. Callbacks MAY decide to consume only one or none of the provided parameters. `when` MUST NOT return a value.
```

Comment:
> Callbacks MAY decide to consume only one or none of the provided parameters.

Not sure if that shall be mentioned in the spec. This is PHP semantics and the spec is unable to enforce or not enforce anything regarding callee behavior here.